lib/active_merchant/billing/gateways/authorize_net.rb

```diff
@@ -305,54 +420,63 @@ def add_check(xml, check)
+          first_name, last_name = names_from(payment_source, address, options)
           xml.firstName(truncate(first_name, 50)) unless empty?(first_name)
           xml.lastName(truncate(last_name, 50)) unless empty?(last_name)
 
-          xml.company(truncate(billing_address[:company], 50)) unless empty?(billing_address[:company])
-          xml.address(truncate(billing_address[:address1], 60))
-          xml.city(truncate(billing_address[:city], 40))
-          xml.state(empty?(billing_address[:state]) ? 'n/a' : truncate(billing_address[:state], 40))
-          xml.zip(truncate((billing_address[:zip] || options[:zip]), 20))
-          xml.country(truncate(billing_address[:country], 60))
-          xml.phoneNumber(truncate(billing_address[:phone], 25)) unless empty?(billing_address[:phone])
-          xml.faxNumber(truncate(billing_address[:fax], 25)) unless empty?(billing_address[:fax])
-        end
-
-        unless shipping_address.blank?
-          xml.shipTo do
-            (first_name, last_name) = if shipping_address[:name]
-              shipping_address[:name].split
-            else
-              [shipping_address[:first_name], shipping_address[:last_name]]
-            end
-            xml.firstName(truncate(first_name, 50)) unless empty?(first_name)
-            xml.lastName(truncate(last_name, 50)) unless empty?(last_name)
-
-            xml.company(truncate(shipping_address[:company], 50)) unless empty?(shipping_address[:company])
-            xml.address(truncate(shipping_address[:address1], 60))
-            xml.city(truncate(shipping_address[:city], 40))
-            xml.state(truncate(shipping_address[:state], 40))
-            xml.zip(truncate(shipping_address[:zip], 20))
-            xml.country(truncate(shipping_address[:country], 60))
-          end
+          xml.company(truncate(address[:company], 50)) unless empty?(address[:company])
+          xml.address(truncate(address[:address1], 60))
+          xml.city(truncate(address[:city], 40))
+          xml.state(empty?(address[:state]) ? 'n/a' : truncate(address[:state], 40))
+          xml.zip(truncate((address[:zip] || options[:zip]), 20))
+          xml.country(truncate(address[:country], 60))
+          xml.phoneNumber(truncate(address[:phone], 25)) unless empty?(address[:phone])
+          xml.faxNumber(truncate(address[:fax], 25)) unless empty?(address[:fax])
         end
+      end
 
-        xml.customerIP(options[:ip]) unless empty?(options[:ip])
+      def add_shipping_address(xml, options, root_node="shipTo")
+        address = options[:shipping_address] || options[:address]
+        return unless address
 
-        xml.cardholderAuthentication do
-          xml.authenticationIndicator(options[:authentication_indicator])
-          xml.cardholderAuthenticationValue(options[:cardholder_authentication_value])
+        xml.send(root_node) do
+          (first_name, last_name) = if address[:name]
+            address[:name].split
```

Comment:
Good catch.  I think this one has been here for awhile but I'll look at tightening it up.